lib/grandpa/finalization.go

```diff
@@ -0,0 +1,493 @@
+			return
+		}
+	}
+}
+
+func (fh *finalizationHandler) stop() (err error) {
+	fh.servicesLock.Lock()
+	defer fh.servicesLock.Unlock()
+
+	finalizationEngineErrCh := make(chan error)
+	go func() {
+		finalizationEngineErrCh <- fh.finalizationEngine.Stop()
+	}()
+
+	votingRoundErrCh := make(chan error)
+	go func() {
+		votingRoundErrCh <- fh.votingRound.Stop()
+	}()
+
+	finalizationEngErr := <-finalizationEngineErrCh
+	votingRoundErr := <-votingRoundErrCh
+
+	if finalizationEngErr != nil && votingRoundErr != nil {
+		return fmt.Errorf("%w: %s; %s", errServicesStopFailed, finalizationEngErr, votingRoundErr)
+	}
+
+	if finalizationEngErr != nil {
+		return fmt.Errorf("%w: %s", errServicesStopFailed, finalizationEngErr)
+	}
+
+	if votingRoundErr != nil {
+		return fmt.Errorf("%w: %s", errServicesStopFailed, votingRoundErr)
+	}
+
+	return nil
+}
+
+func (fh *finalizationHandler) Stop() (err error) {
+	// even though we close the `stopCh` we might
+	// we drain the `errorCh` since an ephemeral service
+	// could failed in the mean time the caller calls the
+	// Stop method hanging here as it will blocks to write
+	// into the channel since there is no channel reader
+	close(fh.stopCh)
+	<-fh.errorCh
+	<-fh.handlerDone
+
+	err = fh.stop()
+	return err
+}
+
+// waitServices will start the ephemeral services that handles the
+// votes for the current round and once those services finishes the
+// waitServices return without errors, if one ephemeral service face
+// a problem the waitServices will shut down all the running ephemeral
+// services and return an error, this function also returns if the
+// finalizationHandler.Stop() method is called
+func (fh *finalizationHandler) waitServices() error {
+	fh.servicesLock.Lock()
+	fh.finalizationEngine, fh.votingRound = fh.newServices()
+	fh.servicesLock.Unlock()
+
+	finalizationEngineErr := make(chan error)
+	go func() {
+		finalizationEngineErr <- fh.finalizationEngine.Run()
+	}()
+
+	votingRoundErr := make(chan error)
+	go func() {
+		votingRoundErr <- fh.votingRound.Run()
+	}()
+
+	for {
+		select {
+		case <-fh.stopCh:
+			return nil
+
+		case err := <-votingRoundErr:
+			if err == nil {
+				votingRoundErr = nil
+				// go out from the select case
+				break
+			}
+
+			stopErr := fh.stop()
+			if stopErr != nil {
+				logger.Warnf("stopping finalisation handler: %s", stopErr)
+			}
+			return err
+
+		case err := <-finalizationEngineErr:
+			if err == nil {
+				finalizationEngineErr = nil
+				// go out from the select case
+				break
+			}
+
+			stopErr := fh.stop()
+			if stopErr != nil {
+				logger.Warnf("stopping finalisation handler: %s", stopErr)
+			}
+			return err
+		}
+
+		finish := votingRoundErr == nil && finalizationEngineErr == nil
+		if finish {
+			return nil
+		}
+	}
+}
+
+// handleVotingRound interacts with finalizationEngine service
+// executing the actions based on what it receives throuhg channel
+type handleVotingRound struct {
```

Comment:
Dumb question, why do we have `handleVotingRound`? It looks like it reads only from `h.finalizationEngineCh` (which is unbuffered)... couldn't we just do this directly in the `finalizationEngine`?

Reply:
The name `handleVotingRound` could be better but the idea is to separate the finalization engine (which count pre-votes, pre-commits, check the round is completable) from the part where we send pre-votes, pre-commits, primary proposals which we call `handleVotingRound`. 

You're right, right now it just performs actions based on what the finalization engine sends but the idea is to evolve to choose the peers to send messages for example. So the separation is intentional.

Reply:
Let's make it without a channel then. Just have it as a method call and inject `handleVotingRound` to the finalization engine. Let's very much 100% avoid channels if unnecessary (for complexity, readability and performance)

Reply:
> Let's make it without a channel then.
@EclesioMeloJunior said the separation is intentional. 🤔

Reply:
After some pairing with Eclesio, I think we should inject the voting round handler to the finalization engine, and have an `enqueueAction(action)` method on the voting round. That way the voting round can handle internally how to enqueue the action (without a channel, with an unbuffered or buffered channel or some other clever queuing system). We can keep the two voting round handler and finalization engine as two goroutines running in parallel.